patterns/trackers.md

```diff
@@ -0,0 +1,56 @@
+---
+title: trackers
+tags:
+  - pattern
+  - protocol
+  - infrastructure
+layout: pattern
+---
+
+## The Design Problem
+
+A common problem in peer-to-peer systems is how to "join" a network. Once a user has found peers offering relevant content they can ask those peers for contact information for adjacent peers, and connect to a larger portion of the network - but finding the initial "introduction" peers is a complicated challenge.
+
+Even once relevant peers are identified, creating a peer-to-peer network connection between two home networks requires NAT traversal, also referred to as NAT hole punching. This usually involves utilizing a third party to exchange network address and port information to facilitate a direct connection.
+
+These challenges do not exist in centralized platforms: On centralized social networks users must identify content and peers relevant to their own interests and "follow" or "subscribe" to the content, but the platform provides the supporting infrastructure to connect with peers and browse available content.
+
+## The Design Solution
+
+Create centralized but redundant "tracker" servers to introduce peers. These trackers have well known addresses so they can be easily found by users, and may even have their addresses hardcoded into client software to aid peer discovery.
+
+Trackers serve as a key/value store, where the "key" is a unique identifier representing content, and the "value" is a list of contact information for introductory peers related to the content. Traditionally this contact information is in the form of an IP address and port number, and may contain additional metadata about the peer's bandwidth and supported functionality. However, the contact information could take a variety of other forms, including onion addresses, public keys, or other content identifiers to indicate related content.
+
+### Examples
+
+Trackers are famously used by torrents to introduce peers to content seeds. They are also used to add new peers to distributed hash tables, or similar communication networks like blockchains. Examples include:
+
+- Torrents
+- Kademlia DHT
+- Blockchains
+- Tor Directory Servers
+- IPFS Bootstrap Peers
+- Cabal
+
+## Why Choose Trackers?
+
+Trackers are a simple technique to solve "first introduction" problems in peer-to-peer systems, provide a natural solution to NAT traversal, and can serve as a bootstrapping step to add clients to a more decentralized network like a distributed hash table.
+
+## Best Practice: How to Implement Trackers
+
+- Include multiple redundant trackers in content links (such as torrent files or magnet links) or hardcoded in client software
```

Comment:
I think it might be unclear to the reader what you mean here, I wonder if it could be explained in more simple terms? Part of the confusion may be that 'content links' isn't really defined here.

Reply:
You're absolutely right. I wanted an abstraction over "torrent files, magnet links, or similar", but stumbled over the phrasing. I want to convey that multiple trackers should be included when the user is given a reference to the content, but most of this pattern takes place once the user already has such a reference.